listenbrainz/spotify_updater/spotify_read_listens.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/python3
+import werkzeug.exceptions
+from dateutil import parser
+import listenbrainz.webserver
+from listenbrainz.domain import spotify
+from listenbrainz.webserver.views.api_tools import insert_payload, validate_listen, LISTEN_TYPE_IMPORT
+from listenbrainz.db import user as db_user
+
+
+def _convert_spotify_play_to_listen(play):
+    track = play['track']
+    album = track['album']
+    # Spotify provides microseconds, but we only give seconds to listenbrainz
+    listened_at = int(parser.parse(play['played_at']).timestamp())
+
+    artists = track['artists']
+    album_artists = album['artists']
+    artist_name = ', '.join([a['name'] for a in artists])
+    album_artist_name = ', '.join([a['name'] for a in album_artists])
+
+    additional = {'tracknumber': track['track_number'],
+                  'spotify_artist_ids': [a['external_urls']['spotify'] for a in artists],
+                  'artist_names': [a['name'] for a in artists],
+                  'listening_from': 'spotify',
+                  'disc_number': track['disc_number'],
+                  'duration_ms': track['duration_ms'],
+                  'spotify_album_id': track['album']['external_urls']['spotify'],
+                  # Named 'release_*' because 'release_name' is an official name in the docs
+                  'release_artist_name': album_artist_name,
+                  'release_artist_names': [a['name'] for a in album_artists],
+                  # Named 'album_*' because Spotify calls it album and this is spotify-specific
+                  'spotify_album_artist_ids': [a['external_urls']['spotify'] for a in album_artists],
+                  }
+    isrc = track.get('external_ids', {}).get('isrc')
+    spotify_url = track.get('external_urls', {}).get('spotify')
+    if isrc:
+        additional['isrc'] = isrc
+    if spotify_url:
+        additional['spotify_id'] = spotify_url
+
+    track_meta = {'artist_name': artist_name,
+                  'track_name': track['name'],
+                  'release_name': album['name'],
+                  'additional_info': additional
+                  }
+
+    return {'listened_at': listened_at,
+            'track_metadata': track_meta}
+
+
+def process_one_user(user):
+    if user.token_expired:
+        spotify.refresh_user_token(user)
+        user = spotify.get_user(user.user_id)
+
+    listenbrainz_user = db_user.get(user.user_id)
+
+    spotipy_client = user.get_spotipy_client()
+
+    recently_played = spotipy_client._get("me/player/recently-played", limit=50)
+    # TODO: if error
+    # If it's 401, try and refresh one more time and get it
+    # If that time fails, report the failure
+    # If refresh fails, report the failure
+    # What other types of error can we make? (Try bad/invalid keys), 404?
+    # If you get status code 429, it means that you have sent too many requests. If this happens, have a look in the Retry-After header
```

Comment:
Spotipy is nice enough to do this for us, it seems. :)

https://github.com/plamere/spotipy/blob/master/spotipy/client.py#L152